From 209d2e4d1912655855e576972fe14d225b2545de Mon Sep 17 00:00:00 2001
From: kitakar5525 <34676735+kitakar5525@users.noreply.github.com>
Date: Tue, 10 Sep 2019 16:28:57 +0900
Subject: [PATCH] ipts: fix crash caused by calling ipts_send_feedback()
 repeatedly
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Calling the "ipts_send_feedback" function repeatedly seems to be
what is causing touch to crash (found by sebanc, see the link below for
the comment) on some models, especially on Surface Pro 4 and
Surface Book 1.
The most desirable fix could be done by raising IPTS GuC priority. Until
we find a better solution, use this workaround.

This commit will disable sending feedback on SP4/SB1 for now.

- SP4/SB1 users can still forcibly enable sending feedback by `intel_ipts.no_feedback=0`
- Users other than SP4/SB1 can forcibly disable sending by `intel_ipts.no_feedback=1`
- Or you can switch it at runtime by `sudo su -c "echo 0 > /sys/module/intel_ipts/parameters/no_feedback`

Link to the comment where sebanc found this workaround:
https://github.com/jakeday/linux-surface/issues/374#issuecomment-508234110
(Touch and pen issue persists · Issue #374 · jakeday/linux-surface)

Link to usage from kitakar5525 who made this commit:
https://github.com/jakeday/linux-surface/issues/374#issuecomment-517289171
(Touch and pen issue persists · Issue #374 · jakeday/linux-surface)

---

changes in v2:
- use DMI matching (currently SB1/SP4)
- added module parameter (no_feedback)
---
 drivers/misc/ipts/ipts-hid.c | 48 ++++++++++++++++++++++++++++++++++++
 1 file changed, 48 insertions(+)

diff --git a/drivers/misc/ipts/ipts-hid.c b/drivers/misc/ipts/ipts-hid.c
index 7c853c1b0..ba8020a96 100644
--- a/drivers/misc/ipts/ipts-hid.c
+++ b/drivers/misc/ipts/ipts-hid.c
@@ -17,6 +17,7 @@
 #include <linux/firmware.h>
 #include <linux/hid.h>
 #include <linux/vmalloc.h>
+#include <linux/dmi.h>
 
 #include "ipts.h"
 #include "ipts-resource.h"
@@ -52,6 +53,30 @@ typedef struct kernel_output_payload_error {
 	char string[128];
 } kernel_output_payload_error_t;
 
+static const struct dmi_system_id no_feedback_dmi_table[] = {
+	{
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Book"),
+		},
+	},
+	{
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 4"),
+		},
+	},
+	{ }
+};
+
+int no_feedback = -1;
+module_param(no_feedback, int, 0644);
+MODULE_PARM_DESC(no_feedback,
+				"Disable sending feedback in order to work around the issue that IPTS "
+				"stops working after some amount of use. "
+				"-1=auto (true if your model is SB1/SP4, false if another model), "
+				"0=false, 1=true, (default: -1)");
+
 static int ipts_hid_get_hid_descriptor(ipts_info_t *ipts, u8 **desc, int *size)
 {
 	u8 *buf;
@@ -415,7 +440,30 @@ static int handle_outputs(ipts_info_t *ipts, int parallel_idx)
 		}
 	}
 
+	/*
+	 * XXX: Calling the "ipts_send_feedback" function repeatedly seems to be
+	 * what is causing touch to crash (found by sebanc, see the link below for
+	 * the comment) on some models, especially on Surface Pro 4 and
+	 * Surface Book 1.
+	 * The most desirable fix could be done by raising IPTS GuC priority. Until
+	 * we find a better solution, use this workaround.
+	 *
+	 * Link to the comment where sebanc found this workaround:
+	 * https://github.com/jakeday/linux-surface/issues/374#issuecomment-508234110
+	 * (Touch and pen issue persists · Issue #374 · jakeday/linux-surface)
+	 *
+	 * Link to the usage from kitakar5525 who made this change:
+	 * https://github.com/jakeday/linux-surface/issues/374#issuecomment-517289171
+	 * (Touch and pen issue persists · Issue #374 · jakeday/linux-surface)
+	 */
 	if (fb_buf) {
+		/* A negative value means "decide by dmi table" */
+		if (no_feedback < 0)
+			no_feedback = dmi_check_system(no_feedback_dmi_table) ? true : false;
+
+		if (no_feedback)
+			return 0;
+
 		ret = ipts_send_feedback(ipts, parallel_idx, transaction_id);
 		if (ret)
 			return ret;
-- 
2.23.0

